src/transformers/tokenization_utils_base.py

```diff
@@ -1241,20 +1274,44 @@ def _from_pretrained(cls, pretrained_model_name_or_path, *init_inputs, **kwargs)
         tokenizer.init_inputs = init_inputs
         tokenizer.init_kwargs = init_kwargs
 
-        # update unique_added_tokens_encoder with special tokens for correct tokenization
-        if hasattr(tokenizer, "unique_added_tokens_encoder"):
-            union = set(tokenizer.unique_added_tokens_encoder).union(tokenizer.all_special_tokens)
-            tokenizer.unique_added_tokens_encoder = list(union)
+        # If there is a complementary special token map, load it
+        special_tokens_map_file = resolved_vocab_files.pop("special_tokens_map_file", None)
+        if special_tokens_map_file is not None:
+            with open(special_tokens_map_file, encoding="utf-8") as special_tokens_map_handle:
+                special_tokens_map = json.load(special_tokens_map_handle)
+
+            for key, value in special_tokens_map.items():
+                if isinstance(value, dict):
+                    value = AddedToken(**value)
+                setattr(tokenizer, key, value)
 
         # Add supplementary tokens.
+        special_tokens = tokenizer.all_special_tokens
         if added_tokens_file is not None:
             with open(added_tokens_file, encoding="utf-8") as added_tokens_handle:
                 added_tok_encoder = json.load(added_tokens_handle)
-            added_tok_decoder = {v: k for k, v in added_tok_encoder.items()}
-            tokenizer.added_tokens_encoder.update(added_tok_encoder)
-            tokenizer.added_tokens_decoder.update(added_tok_decoder)
-            union = set(tokenizer.unique_added_tokens_encoder).union(tokenizer.added_tokens_encoder.keys())
-            tokenizer.unique_added_tokens_encoder = list(union)
+
+            # Sort added tokens by index
+            added_tok_encoder_sorted = list(sorted(added_tok_encoder.items(), key=lambda x: x[1]))
+
+            for token, index in added_tok_encoder_sorted:
+                assert index == len(tokenizer), (
+                    f"Non-consecutive added token '{token}' found. "
+                    f"Should have index {len(tokenizer)} but has index {index} in saved vocabulary."
+                )
```

Comment:
This will now raise an error if non-consecutive tokens are provided in the serialized vocabulary.

I contemplated making this a warning only but I think it's better to enforce good practices here rather than keep backward compatibility. If your vocabulary has "holes" in it something went wrong somewhere and reassigning the token to a new index will be a source of silent errors.

Reply:
Won't this fail with CTRL? I recall hearing that CTRL had such an issue